next/babel preset hard-codes directory path in output

The `babelRuntimePath` at the beginning of the `babel/preset.js` file results in the absolute path to the babel runtime being hard-coded into babel output. This means that e.g. server code cannot be babel-compiled in one location and run on another.

I'm running into this by babel-compiling my server code on one machine and deploying the compiled versions to a server and running them there.

Looks like the webpack config is doing this, too. Here’s what it looks like in Heroku logs:

```
2017-02-15T19:04:09.462454+00:00 app[web.1]: { Error: Cannot find module '/tmp/build_c63e484ca4c05f0e45d82a9e42201fc2/node_modules/next/dist/lib/head.js'
2017-02-15T19:04:09.462459+00:00 app[web.1]:     at Function.Module._resolveFilename (module.js:470:15)
2017-02-15T19:04:09.462460+00:00 app[web.1]:     at Function.Module._load (module.js:418:25)
2017-02-15T19:04:09.462461+00:00 app[web.1]:     at Module.require (module.js:498:17)
2017-02-15T19:04:09.462461+00:00 app[web.1]:     at require (internal/module.js:20:19)
2017-02-15T19:04:09.462462+00:00 app[web.1]:     at Object.<anonymous> (/app/.next/dist/pages/report.js:39:13)
2017-02-15T19:04:09.462463+00:00 app[web.1]:     at Module._compile (module.js:571:32)
2017-02-15T19:04:09.462463+00:00 app[web.1]:     at Object.Module._extensions..js (module.js:580:10)
2017-02-15T19:04:09.462464+00:00 app[web.1]:     at Module.load (module.js:488:32)
2017-02-15T19:04:09.462464+00:00 app[web.1]:     at tryModuleLoad (module.js:447:12)
2017-02-15T19:04:09.462465+00:00 app[web.1]:     at Function.Module._load (module.js:439:3) code: 'MODULE_NOT_FOUND' }
```

Note how Heroku built in a tmp directory but is running from an /app directory.

I can have a PR for this up pretty soon.

--- a/server/build/babel/preset.js
+++ b/server/build/babel/preset.js
@@ -1,5 +1,4 @@
-const babelRuntimePath = require.resolve('babel-runtime/package')
-  .replace(/[\\/]package\.json$/, '')
+const relativeResolve = require('../root-module-relative-path').default(require)
 
 const envPlugins = {
   'development': [
@@ -30,14 +29,14 @@ module.exports = {
       require.resolve('babel-plugin-module-resolver'),
       {
         alias: {
-          'babel-runtime': babelRuntimePath,
-          'next/link': require.resolve('../../../lib/link'),
-          'next/prefetch': require.resolve('../../../lib/prefetch'),
-          'next/css': require.resolve('../../../lib/css'),
-          'next/head': require.resolve('../../../lib/head'),
-          'next/document': require.resolve('../../../server/document'),
-          'next/router': require.resolve('../../../lib/router'),
-          'next/error': require.resolve('../../../lib/error')
+          'babel-runtime': relativeResolve('babel-runtime/package'),
+          'next/link': relativeResolve('../../../lib/link'),
+          'next/prefetch': relativeResolve('../../../lib/prefetch'),
+          'next/css': relativeResolve('../../../lib/css'),
+          'next/head': relativeResolve('../../../lib/head'),
+          'next/document': relativeResolve('../../../server/document'),
+          'next/router': relativeResolve('../../../lib/router'),
+          'next/error': relativeResolve('../../../lib/error')
         }
       }
     ]

--- a/server/build/root-module-relative-path.js
+++ b/server/build/root-module-relative-path.js
@@ -1,0 +1,24 @@
+// Next.js needs to use module.resolve to generate paths to modules it includes,
+// but those paths need to be relative to something so that they're portable
+// across directories and machines.
+//
+// This function returns paths relative to the top-level 'node_modules'
+// directory found in the path. If none is found, returns the complete path.
+
+const RELATIVE_START = 'node_modules/'
+
+// Pass in the module's `require` object since it's module-specific.
+export default (moduleRequire) => (path) => {
+  // package.json removed because babel-runtime is resolved as
+  // "babel-runtime/package"
+  const absolutePath = moduleRequire.resolve(path)
+    .replace(/[\\/]package\.json$/, '')
+
+  const relativeStartIndex = absolutePath.indexOf(RELATIVE_START)
+
+  if (relativeStartIndex === -1) {
+    return absolutePath
+  }
+
+  return absolutePath.substring(relativeStartIndex + RELATIVE_START.length)
+}

--- a/server/build/webpack.js
+++ b/server/build/webpack.js
@@ -11,6 +11,7 @@ import JsonPagesPlugin from './plugins/json-pages-plugin'
 import getConfig from '../config'
 import * as babelCore from 'babel-core'
 import findBabelConfig from './babel/find-config'
+import rootModuleRelativePath from './root-module-relative-path'
 
 const documentPage = join('pages', '_document.js')
 const defaultPages = [
@@ -22,6 +23,8 @@ const nextNodeModulesDir = join(__dirname, '..', '..', '..', 'node_modules')
 const interpolateNames = new Map(defaultPages.map((p) => {
   return [join(nextPagesDir, p), `dist/pages/${p}`]
 }))
+
+const relativeResolve = rootModuleRelativePath(require)
 
 export default async function createCompiler (dir, { dev = false, quiet = false, buildDir } = {}) {
   dir = resolve(dir)
@@ -164,8 +167,7 @@ export default async function createCompiler (dir, { dev = false, quiet = false,
         if (!(/\.js$/.test(interpolatedName))) {
           return { content, sourceMap }
         }
-        const babelRuntimePath = require.resolve('babel-runtime/package')
-          .replace(/[\\/]package\.json$/, '')
+
         const transpiled = babelCore.transform(content, {
           babelrc: false,
           sourceMaps: dev ? 'both' : false,
@@ -180,15 +182,15 @@ export default async function createCompiler (dir, { dev = false, quiet = false,
               require.resolve('babel-plugin-module-resolver'),
               {
                 alias: {
-                  'babel-runtime': babelRuntimePath,
-                  'next/link': require.resolve('../../lib/link'),
-                  'next/prefetch': require.resolve('../../lib/prefetch'),
-                  'next/css': require.resolve('../../lib/css'),
-                  'next/head': require.resolve('../../lib/head'),
-                  'next/document': require.resolve('../../server/document'),
-                  'next/router': require.resolve('../../lib/router'),
-                  'next/error': require.resolve('../../lib/error'),
-                  'styled-jsx/style': require.resolve('styled-jsx/style')
+                  'babel-runtime': relativeResolve('babel-runtime/package'),
+                  'next/link': relativeResolve('../../lib/link'),
+                  'next/prefetch': relativeResolve('../../lib/prefetch'),
+                  'next/css': relativeResolve('../../lib/css'),
+                  'next/head': relativeResolve('../../lib/head'),
+                  'next/document': relativeResolve('../../server/document'),
+                  'next/router': relativeResolve('../../lib/router'),
+                  'next/error': relativeResolve('../../lib/error'),
+                  'styled-jsx/style': relativeResolve('styled-jsx/style')
                 }
               }
             ]

